packages/backend-defaults/src/entrypoints/actions/DefaultActionsService.ts

```diff
@@ -0,0 +1,137 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  ActionsService,
+  ActionsServiceAction,
+  AuthService,
+  BackstageCredentials,
+  DiscoveryService,
+  LoggerService,
+  RootConfigService,
+} from '@backstage/backend-plugin-api';
+import { ResponseError } from '@backstage/errors';
+import { JsonObject } from '@backstage/types';
+
+export class DefaultActionsService implements ActionsService {
+  private constructor(
+    private readonly discovery: DiscoveryService,
+    private readonly config: RootConfigService,
+    private readonly logger: LoggerService,
+    private readonly auth: AuthService,
+  ) {}
+
+  static create({
+    discovery,
+    config,
+    logger,
+    auth,
+  }: {
+    discovery: DiscoveryService;
+    config: RootConfigService;
+    logger: LoggerService;
+    auth: AuthService;
+  }) {
+    return new DefaultActionsService(discovery, config, logger, auth);
+  }
+
+  async listActions({ credentials }: { credentials: BackstageCredentials }) {
+    const pluginSources =
+      this.config.getOptionalStringArray('backend.actions.pluginSources') ?? [];
+
+    const remoteActionsList = await Promise.all(
+      pluginSources.map(async source => {
+        const pluginBaseUrl = await this.discovery.getBaseUrl(source);
+        const response = await this.makeRequest({
+          url: `${pluginBaseUrl}/.backstage/actions/v1/actions`,
+          pluginId: source,
+          credentials,
+        });
+
+        if (!response.ok) {
+          this.logger.warn(`Failed to fetch actions from ${source}`);
+          return [];
+        }
+
+        const { actions } = (await response.json()) as {
+          actions: ActionsServiceAction;
+        };
+
+        return actions;
+      }),
+    );
+
+    return { actions: remoteActionsList.flat() };
+  }
+
+  async invokeAction(opts: {
+    id: string;
+    input?: JsonObject;
+    credentials: BackstageCredentials;
+  }) {
+    const pluginId = this.pluginIdFromActionId(opts.id);
+
+    const baseUrl = await this.discovery.getBaseUrl(pluginId);
+
+    const response = await this.makeRequest({
+      url: `${baseUrl}/.backstage/actions/v1/actions/${opts.id}/invoke`,
+      pluginId,
+      credentials: opts.credentials,
+      options: {
+        method: 'POST',
+        body: JSON.stringify(opts.input),
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      },
+    });
+
+    if (!response.ok) {
+      throw await ResponseError.fromResponse(response);
+    }
+
+    const { output } = await response.json();
+    return { output };
+  }
+
+  private async makeRequest(opts: {
```

Comment:
You could even make the `getBaseUrl` bit part of this too and make `url` the path